File: projects/03-realtime-pubsub/src/presence.rs
//! V3 — Presence: who is currently in each topic.
//!
//! Presence is **soft state** — only ever an approximation of reality that must
//! converge as connections come and go. The [`PresenceRegistry`] tracks, per
//! topic, which connections are present and the display identity each one
//! claimed.
//!
//! The easy half is the clean lifecycle: [`PresenceRegistry::join`] on
//! subscribe, [`PresenceRegistry::leave`] on unsubscribe. The hard half is
//! *absence*: a client whose network drops never sends a leave.
//! [`PresenceRegistry::disconnect`] covers every observed teardown path;
//! [`PresenceRegistry::touch`] plus [`PresenceRegistry::sweep`] handle silent
//! vanish via heartbeat + TTL.

use std::{
    collections::HashMap,
    time::{Duration, Instant},
};

use parking_lot::RwLock;

use crate::protocol::{ConnId, Topic};

/// One connection's membership in a topic: display name plus liveness stamp.
///
/// `last_seen` is refreshed by [`Member::refresh`] (via
/// [`PresenceRegistry::touch`]) so [`PresenceRegistry::sweep`] can expire
/// entries that stop heartbeating.
#[derive(Clone, Debug, PartialEq, Eq)]
pub struct Member {
    identity: String,
    last_seen: Instant,
}

impl Member {
    /// Create a member with `identity` and `last_seen` set to now.
    pub fn new(identity: String) -> Self {
        Self {
            identity,
            last_seen: Instant::now(),
        }
    }

    /// Display identity claimed for this membership (wire-facing name).
    pub fn identity(&self) -> &str {
        &self.identity
    }

    /// Mark this member as seen now (heartbeat / any liveness proof).
    pub fn refresh(&mut self) {
        self.last_seen = Instant::now();
    }

    #[cfg(test)]
    fn age_for_test(&mut self, by: Duration) {
        self.last_seen -= by;
    }
}

/// Per-topic presence registry: who is currently in each room.
///
/// Orthogonal to the fan-out [`crate::hub::Hub`]: the hub answers "who receives
/// messages?", this answers "who appears in the room list?". Interior-mutable
/// (`&self` methods) so it can live behind an `Arc` in `AppState`.
#[derive(Default)]
pub struct PresenceRegistry {
    topic_members: RwLock<HashMap<Topic, HashMap<ConnId, Member>>>,
}

impl PresenceRegistry {
    /// Create an empty registry with no topics and no members.
    pub fn new() -> Self {
        Self::default()
    }

    /// Record `conn` as present in `topic` under `identity`.
    ///
    /// Re-joining the same `conn` replaces the previous [`Member`] (new
    /// identity and a fresh `last_seen`). Typical call site: subscribe.
    pub fn join(&self, topic: &str, conn: ConnId, identity: String) {
        let mut members = self.topic_members.write();
        members
            .entry(topic.to_string())
            .or_default()
            .insert(conn, Member::new(identity));
    }

    /// Remove `conn` from `topic`, pruning the topic when it becomes empty.
    ///
    /// Idempotent: a missing topic or unknown `conn` is a no-op aside from
    /// briefly ensuring the topic entry exists before removal.
    pub fn leave(&self, topic: &str, conn: ConnId) {
        let mut members = self.topic_members.write();
        members.entry(topic.to_string()).or_default().remove(&conn);
        if members.get(topic).unwrap().is_empty() {
            members.remove(topic);
        }
    }

    /// Snapshot the members currently present in `topic`.
    ///
    /// Returns an empty vec when the topic is unknown or has been pruned —
    /// both mean "nobody here" for soft-state presence.
    pub fn members(&self, topic: &str) -> Vec<Member> {
        let members = self.topic_members.read();
        if members.contains_key(topic) {
            members.get(topic).unwrap().values().cloned().collect()
        } else {
            Vec::new()
        }
    }

    /// Remove `conn` from every topic it appears in, pruning empty topics.
    ///
    /// The anti-ghost catch-all: call on every WebSocket teardown path (clean
    /// close, error, abrupt drop). Idempotent if `conn` was never present.
    pub fn disconnect(&self, conn: ConnId) {
        let mut members = self.topic_members.write();
        let mut keys_to_remove = Vec::new();
        for (topic, members) in members.iter_mut() {
            members.remove(&conn);
            if members.is_empty() {
                keys_to_remove.push(topic.clone());
            }
        }
        for topic in keys_to_remove {
            members.remove(&topic);
        }
    }

    /// Refresh `last_seen` for `conn` in every topic it belongs to.
    ///
    /// No-op if `conn` is not present anywhere. Wire this from WebSocket
    /// ping/pong (or any client traffic) so live sockets stay under the TTL.
    pub fn touch(&self, conn: ConnId) {
        let mut members = self.topic_members.write();
        for (_, members) in members.iter_mut() {
            if let Some(member) = members.get_mut(&conn) {
                member.refresh();
            }
        }
    }

    #[cfg(test)]
    fn age(&self, conn: ConnId, by: Duration) {
        let mut members = self.topic_members.write();
        for topic_members in members.values_mut() {
            if let Some(member) = topic_members.get_mut(&conn) {
                member.age_for_test(by);
            }
        }
    }

    /// Evict members whose `last_seen` is at least `ttl` old.
    ///
    /// Returns only topics that lost at least one member, each paired with the
    /// **remaining** members after eviction (empty if the room was fully
    /// cleared). Unchanged topics are omitted so a background task can
    /// broadcast [`crate::protocol::ServerMessage::Presence`] only when needed.
    ///
    /// Intended to be called on an interval from a `tokio` task; this method
    /// itself is synchronous and lock-scoped.
    pub fn sweep(&self, ttl: Duration) -> Vec<(Topic, Vec<Member>)> {
        let mut topic_members = self.topic_members.write();
        let mut changed = Vec::new();

        topic_members.retain(|topic, members| {
            let before = members.len();
            members.retain(|_, member| member.last_seen.elapsed() < ttl);
            if members.len() != before {
                changed.push((topic.clone(), members.values().cloned().collect()));
            }
            !members.is_empty()
        });

        changed
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    fn identities(members: Vec<Member>) -> Vec<String> {
        let mut ids: Vec<String> = members
            .into_iter()
            .map(|m| m.identity().to_string())
            .collect();
        ids.sort();
        ids
    }

    #[test]
    fn join_two_conns_returns_both_identities() {
        let presence = PresenceRegistry::new();
        let a = ConnId::next();
        let b = ConnId::next();

        presence.join("room1", a, "alice".into());
        presence.join("room1", b, "bob".into());

        assert_eq!(
            identities(presence.members("room1")),
            vec!["alice".to_string(), "bob".to_string()]
        );
    }

    #[test]
    fn leave_one_leaves_the_other() {
        let presence = PresenceRegistry::new();
        let a = ConnId::next();
        let b = ConnId::next();

        presence.join("room1", a, "alice".into());
        presence.join("room1", b, "bob".into());
        presence.leave("room1", a);

        assert_eq!(
            identities(presence.members("room1")),
            vec!["bob".to_string()]
        );
    }

    #[test]
    fn leave_last_member_prunes_topic() {
        let presence = PresenceRegistry::new();
        let a = ConnId::next();

        presence.join("room1", a, "alice".into());
        presence.leave("room1", a);

        assert!(presence.members("room1").is_empty());
    }

    #[test]
    fn disconnect_removes_conn_from_every_topic() {
        let presence = PresenceRegistry::new();
        let a = ConnId::next();
        let b = ConnId::next();

        presence.join("room1", a, "alice".into());
        presence.join("room2", a, "alice".into());
        presence.join("room1", b, "bob".into());

        presence.disconnect(a);

        assert_eq!(
            identities(presence.members("room1")),
            vec!["bob".to_string()]
        );
        assert!(presence.members("room2").is_empty());
    }

    #[test]
    fn members_on_unknown_topic_is_empty() {
        let presence = PresenceRegistry::new();
        assert!(presence.members("nobody-here").is_empty());
    }

    #[test]
    fn join_same_conn_refreshes_identity() {
        let presence = PresenceRegistry::new();
        let a = ConnId::next();

        presence.join("room1", a, "alice".into());
        presence.join("room1", a, "alice-v2".into());

        assert_eq!(
            identities(presence.members("room1")),
            vec!["alice-v2".to_string()]
        );
    }

    #[test]
    fn sweep_returns_nothing_when_everyone_is_fresh() {
        let presence = PresenceRegistry::new();
        let a = ConnId::next();
        presence.join("room1", a, "alice".into());

        let changed = presence.sweep(Duration::from_secs(30));
        assert!(changed.is_empty());
        assert_eq!(
            identities(presence.members("room1")),
            vec!["alice".to_string()]
        );
    }

    #[test]
    fn sweep_returns_survivors_when_someone_expires() {
        let presence = PresenceRegistry::new();
        let a = ConnId::next();
        let b = ConnId::next();
        presence.join("room1", a, "alice".into());
        presence.join("room1", b, "bob".into());
        presence.age(b, Duration::from_secs(60));

        let changed = presence.sweep(Duration::from_secs(30));
        assert_eq!(changed.len(), 1);
        assert_eq!(changed[0].0, "room1");
        assert_eq!(identities(changed[0].1.clone()), vec!["alice".to_string()]);
        assert_eq!(
            identities(presence.members("room1")),
            vec!["alice".to_string()]
        );
    }

    #[test]
    fn sweep_returns_empty_members_when_room_fully_cleared() {
        let presence = PresenceRegistry::new();
        let a = ConnId::next();
        presence.join("room1", a, "alice".into());
        presence.age(a, Duration::from_secs(60));

        let changed = presence.sweep(Duration::from_secs(30));
        assert_eq!(changed.len(), 1);
        assert_eq!(changed[0].0, "room1");
        assert!(changed[0].1.is_empty());
        assert!(presence.members("room1").is_empty());
    }

    #[test]
    fn touch_keeps_member_from_being_swept() {
        let presence = PresenceRegistry::new();
        let a = ConnId::next();
        presence.join("room1", a, "alice".into());
        presence.age(a, Duration::from_secs(60));
        presence.touch(a);

        let changed = presence.sweep(Duration::from_secs(30));
        assert!(changed.is_empty());
        assert_eq!(
            identities(presence.members("room1")),
            vec!["alice".to_string()]
        );
    }

    #[test]
    fn touch_unknown_conn_is_a_noop() {
        let presence = PresenceRegistry::new();
        let a = ConnId::next();
        let ghost = ConnId::next();
        presence.join("room1", a, "alice".into());

        presence.touch(ghost);

        assert_eq!(
            identities(presence.members("room1")),
            vec!["alice".to_string()]
        );
        assert!(presence.sweep(Duration::from_secs(30)).is_empty());
    }

    #[test]
    fn touch_refreshes_conn_in_every_topic() {
        let presence = PresenceRegistry::new();
        let a = ConnId::next();
        presence.join("room1", a, "alice".into());
        presence.join("room2", a, "alice".into());
        presence.age(a, Duration::from_secs(60));
        presence.touch(a);

        assert!(presence.sweep(Duration::from_secs(30)).is_empty());
        assert_eq!(
            identities(presence.members("room1")),
            vec!["alice".to_string()]
        );
        assert_eq!(
            identities(presence.members("room2")),
            vec!["alice".to_string()]
        );
    }

    #[test]
    fn touch_only_refreshes_the_targeted_conn() {
        let presence = PresenceRegistry::new();
        let a = ConnId::next();
        let b = ConnId::next();
        presence.join("room1", a, "alice".into());
        presence.join("room1", b, "bob".into());
        presence.age(a, Duration::from_secs(60));
        presence.age(b, Duration::from_secs(60));
        presence.touch(a);

        let changed = presence.sweep(Duration::from_secs(30));
        assert_eq!(changed.len(), 1);
        assert_eq!(changed[0].0, "room1");
        assert_eq!(identities(changed[0].1.clone()), vec!["alice".to_string()]);
        assert_eq!(
            identities(presence.members("room1")),
            vec!["alice".to_string()]
        );
    }
}

File: projects/03-realtime-pubsub/src/realtime_pubsub/presence.py
"""V3 — Presence: who is currently in each topic.

Presence is **soft state** — only ever an approximation of reality that must
converge as connections come and go. `PresenceRegistry` tracks, per topic, which
connections are present and the display identity each one claimed.

The easy half is the clean lifecycle: `join` on subscribe, `leave` on
unsubscribe. The hard half is *absence*: a client whose laptop lid closes never
sends a leave, and from the server's side "gone" and "quiet" look identical.
`disconnect` covers every teardown path the server actually observes; `touch`
plus `sweep` handle the silent vanish via heartbeat + TTL.

## The clock

`last_seen` is `time.monotonic()`, not `time.time()`. A TTL measured against the
wall clock is a TTL that can be broken by NTP: a clock that steps backwards
makes every member look freshly-seen and the sweep stops evicting; a step
forward evicts a room full of live people at once. Monotonic time only ever
moves forward and is immune to both. The trade-off — the value is meaningless
across processes and unusable as a timestamp — costs nothing here, because
nothing outside this module ever reads it.
"""

from __future__ import annotations

import time
from dataclasses import dataclass, field

from .protocol import ConnId, Topic

__all__ = ["Member", "PresenceRegistry"]


@dataclass(slots=True)
class Member:
    """One connection's membership in a topic: display name plus liveness stamp."""

    identity: str
    """Display identity claimed for this membership. Client-supplied, so it is
    **never** trusted for anything but display (SPEC security checklist)."""

    last_seen: float = field(default_factory=time.monotonic)
    """Monotonic seconds. Refreshed by `PresenceRegistry.touch`; compared
    against the TTL by `sweep`."""

    def refresh(self) -> None:
        """Mark this member as seen now (heartbeat or any other liveness proof)."""
        self.last_seen = time.monotonic()


class PresenceRegistry:
    """Per-topic presence: who is currently in each room.

    Orthogonal to `Hub`: the hub answers "who receives messages?", this answers
    "who appears in the room list?". Two connections can be in one and not the
    other — a socket that subscribed but whose identity was rejected, or a
    member whose TTL expired while its socket is still technically open.

    Like the hub, every method here is synchronous and `await`-free, so the map
    is never observed half-mutated.
    """

    __slots__ = ("_topic_members",)

    def __init__(self) -> None:
        self._topic_members: dict[Topic, dict[ConnId, Member]] = {}

    def join(self, topic: str, conn: ConnId, identity: str) -> None:
        """Record `conn` as present in `topic` under `identity`.

        Re-joining replaces the previous member (new identity, fresh stamp).
        """
        self._topic_members.setdefault(topic, {})[conn] = Member(identity)

    def leave(self, topic: str, conn: ConnId) -> None:
        """Remove `conn` from `topic`, pruning the topic when it empties.

        Idempotent — an unknown topic or connection is a no-op.
        """
        members = self._topic_members.get(topic)
        if members is None:
            return
        members.pop(conn, None)
        if not members:
            del self._topic_members[topic]

    def members(self, topic: str) -> list[Member]:
        """Snapshot the members currently present in `topic`.

        Empty for an unknown or pruned topic — both mean "nobody here" to
        soft-state presence, and collapsing them keeps callers from having to
        distinguish a room that never existed from one everybody left.
        """
        return list(self._topic_members.get(topic, {}).values())

    def identities(self, topic: str) -> list[str]:
        """Just the display names — what a `presence` frame carries."""
        return [member.identity for member in self._topic_members.get(topic, {}).values()]

    def disconnect(self, conn: ConnId) -> None:
        """Remove `conn` from every topic it appears in, pruning empty topics.

        The anti-ghost catch-all: call on every WebSocket teardown path (clean
        close, protocol error, abrupt drop, server shutdown). Idempotent if
        `conn` was never present.
        """
        for topic in list(self._topic_members):
            members = self._topic_members[topic]
            members.pop(conn, None)
            if not members:
                del self._topic_members[topic]

    def touch(self, conn: ConnId) -> None:
        """Refresh `last_seen` for `conn` in every topic it belongs to.

        No-op if `conn` is not present anywhere. Wired from the client's
        `heartbeat` frame — and worth wiring from *any* traffic that connection
        sends, since a client that is publishing is self-evidently alive.
        """
        for members in self._topic_members.values():
            member = members.get(conn)
            if member is not None:
                member.refresh()

    def sweep(self, ttl: float) -> list[tuple[Topic, list[Member]]]:
        """Evict members whose `last_seen` is at least `ttl` seconds old.

        Returns only the topics that actually lost someone, each paired with the
        **remaining** members (empty if the room cleared out). Unchanged topics
        are omitted so the background task can broadcast a `presence` frame only
        where it would say something new — a sweep that re-broadcast every room
        on every tick would be its own thundering herd.
        """
        cutoff = time.monotonic() - ttl
        changed: list[tuple[Topic, list[Member]]] = []

        for topic in list(self._topic_members):
            members = self._topic_members[topic]
            expired = [conn for conn, member in members.items() if member.last_seen <= cutoff]
            if not expired:
                continue
            for conn in expired:
                del members[conn]
            changed.append((topic, list(members.values())))
            if not members:
                del self._topic_members[topic]

        return changed

    # --- introspection (metrics + tests) ---------------------------------------

    def topics(self) -> list[Topic]:
        """Rooms with at least one member.

        V4 wants this: a node should subscribe to the bus channel for a room
        only while it actually has someone in it.
        """
        return list(self._topic_members)

    def member_count(self) -> int:
        """Total memberships across every room — the `PRESENCE_MEMBERS` gauge."""
        return sum(len(members) for members in self._topic_members.values())
